sdk/resource/resource.go

```diff
@@ -17,30 +17,53 @@
 package resource
 
 import (
-	"reflect"
+	"sort"
+	"strings"
 
 	"go.opentelemetry.io/otel/api/core"
 )
 
 // Resource describes an entity about which identifying information and metadata is exposed.
 type Resource struct {
-	labels map[core.Key]core.Value
+	str    string
+	sorted []core.KeyValue
+	keys   map[core.Key]struct{}
 }
 
 // New creates a resource from a set of attributes.
 // If there are duplicates keys then the first value of the key is preserved.
 func New(kvs ...core.KeyValue) *Resource {
-	res := &Resource{
-		labels: map[core.Key]core.Value{},
-	}
+	res := &Resource{keys: make(map[core.Key]struct{})}
 	for _, kv := range kvs {
-		if _, ok := res.labels[kv.Key]; !ok {
-			res.labels[kv.Key] = kv.Value
+		// First key wins.
+		if _, ok := res.keys[kv.Key]; !ok {
+			res.keys[kv.Key] = struct{}{}
+			res.sorted = append(res.sorted, kv)
 		}
 	}
+	sort.Slice(res.sorted, func(i, j int) bool {
+		return res.sorted[i].Key < res.sorted[j].Key
+	})
+	res.str = buildResourceString(res.sorted)
 	return res
 }
 
+// String implements the Stringer interface and provides a reproducibly
+// hashable representation of a Resource.
+func (r Resource) String() string {
+	return r.str
+}
```

Comment:
This can contain the caching logic if it is changed to a pointer receiver.  It looks like `New()` and `Merge()` both return `*Resource`.

```suggestion
func (r *Resource) String() string {
	if r.str == "" {
		r.str = buildResourceString(r.sorted)
	}
	return r.str
}
```

Reply:
If we decide to do a lazy computation of the string representation, then it needs to sit behind a `sync.Once` object.

Personally I'd lean towards an eager computation, I think.

Reply:
@krnowak: I was adding the early initialization to optimize for the expected calls to this when filtering in the exporter, but, on second thought, I think I'll revert to the eager evaluation to start with. I'm realizing that this is optimizing without a measured need, and this PR was more about the method than the optimized performance here.

@Aneurysm9: I was hoping to eventually change the return values for the `Merge` and `New` functions to be structs instead of pointers to keep things on the stack.